backends/redis.go

```diff
@@ -102,7 +102,7 @@ func (b *RedisBackend) Get(ctx context.Context, key string) (string, error) {
 func (b *RedisBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
 
 	success, err := b.client.Put(ctx, key, value, ttlSeconds)
-	if !success {
+	if err == redis.Nil && !success {
```

Comment:
I started PRebid-Cache and an instance of a Redis Server and noticed that the change proposed in this pull request doesn't have the intended outcome. Suppose we store an XML value under key `"3e6713fb-1aea-4867-be7b-ffc75c8f74dd"`:
```
00:12 $ curl -i -d '{"puts":[{"type":"xml","ttlseconds":60,"key":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd","value":"<?xml version=\"1.0\">__video_info__<\\/VAST>\r\n"}]}' http://localhost:2424/cache
{"responses":[{"uuid":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd"}]}
```
# Before
### Internal server error
If there's an error on the server side the error is successfully displayed, correcting the current error in production:
```
00:12 $ curl -i -d '{"puts":[{"type":"xml","ttlseconds":60,"key":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd","value":"<?xml version=\"1.0\">__video_info__<\\/VAST>\r\n"}]}' http://localhost:2424/cache
HTTP/1.1 500 Internal Server Error
   ...
dial tcp 127.0.0.1:6379: connect: connection refused   # CORRECT, 500 error code
```
### No internal server error, sucessful overwrite:
Unfortunately, with the change proposed in this PR, if we try to overwrite we get don't the desired output, which is to not overwrite the value nor return an error, but simply retern a JSON response with an empty `"uuid"` field:
```
00:12 $ curl -i -d '{"puts":[{"type":"xml","ttlseconds":60,"key":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd","value":"<?xml version=\"1.0\">__video_info__<\\/VAST>\r\n"}]}' http://localhost:2424/cache
HTTP/1.1 200 OK
   ...
{"responses":[{"uuid":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd"}]}  # WRONG, uuid should be blank
```

# After
The following fix gives us the desired solutions on both fronts:
```
102   func (b *RedisBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
103   
104       success, err := b.client.Put(ctx, key, value, ttlSeconds)
105 -     if err == redis.Nil && !success {
    +     if err != redis.Nil {
    +         return err
    +     }
    +     if !success {
106           return utils.NewPBCError(utils.RECORD_EXISTS)
107       }
108 -     return err
108 +     return nil
109   }
backends/redis.go
```
### Internal server error
```
03:05 $ curl -i -d '{"puts":[{"type":"xml","ttlseconds":60,"key":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd","value":"<?xml version=\"1.0\">__video_info__<\\/VAST>\r\n"}]}' http://localhost:2424/cache
HTTP/1.1 500 Internal Server Error
   ...
dial tcp 127.0.0.1:6379: connect: connection refused    ## CORRECT, 500 error code
```
### No internal server error, sucessful overwrite:
```
03:05 $ curl -i -d '{"puts":[{"type":"xml","ttlseconds":60,"key":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd","value":"<?xml version=\"1.0\">__video_info__<\\/VAST>\r\n"}]}' http://localhost:2424/cache
HTTP/1.1 200 OK
   ...
{"responses":[{"uuid":""}]}  # CORRECT no error returned and empty uuid field
```

I'm curious to know your take. Did you get similar outcomes? If you agree, can you please modify along with the test cases?

Reply:
@guscarreon Sorry for the late answer, I've been very busy lately.
I'm a bit surprised by what you observed, since both versions translate to the same result to me.
Let me draft a truth table:
<table>
<tr>
	<td>err
	<td>success
	<td>before returns
	<td>after returns
<tr>
	<td>err
	<td>false
	<td>utils.RECORD_EXISTS
	<td>utils.RECORD_EXISTS
<tr>
	<td>redis.Nil
	<td>true
	<td>err (= redis.Nil)
	<td>nil
<tr>
	<td>!= redis.Nil
	<td>false
	<td>err
	<td>err
<tr>
	<td>!= redis.Nil
	<td>true
	<td>err
	<td>err
</table>

From here, the only difference is that the "after" function returns `nil` instead of `err` (=`redis.Nil`) when everything went well. Just to confirm this, I added an additional test case for { err = redis.Nil, success = true }.

I tried to reproduce your experiment but could not get Prebid Cache to store the value I send it under the designated `key`, so it returns me a different `uuid`` for each attempt and I couldn't get the `RECORD_EXISTS` error.

Did you populate the cache again after each time you shut it down for the failure test ?
If it's not configured to dump its content and it was not re-populated, it's normal that the new attempt to write the key succeeds.

Reply:
Hi @sebmil-daily. I'm sorry for my late response. I see the `if err != nil && err != redis.Nil` check to be common in the `"github.com/go-redis/redis/v8"` repository so I believe we should make that distintion. `redis.Nil` definition is shown below:
``` 
22 const Nil = RedisError("redis: nil") // nolint:errname
23
24 type RedisError string
25
26 func (e RedisError) Error() string { return string(e) }
27
28 func (RedisError) RedisError() {}
29
~/newgo/pkg/mod/github.com/go-redis/redis/v8@v8.11.5/internal/proto/reader.go
```
I'm not sure if there's a scenario where the redis client's `Result() (bool, error)` method throws an `err` equal to `nil`, but I believe it's better to code defensively on it. Can we please modify to account to this scenario? If you agree with the proposal below, I ran the tests again it is working as expected:
```
102   func (b *RedisBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
103   
104       success, err := b.client.Put(ctx, key, value, ttlSeconds)
105 -     if err == redis.Nil && !success {
    +     if err != redis.Nil && err != nil {
    +         return err
    +     }
    +     if !success {
106           return utils.NewPBCError(utils.RECORD_EXISTS)
107       }
108 -     return err
108 +     return nil
109   }
backends/redis.go
```

> tried to reproduce your experiment but could not get Prebid Cache to store the value I send it under the designated `key`, so it returns me a different `uuid` for each attempt and I couldn't get the `RECORD_EXISTS` error. 



In order to store a value under a custom/designated `key`, Prebid Cache needs to be configured with the `request_limits.allow_setting_keys` flag set to `true`:
```
 request_limits:
   allow_setting_keys: true
config.yaml
```
Then passs the custom key in the PUT request body under the `"key"` field: `curl -i -d '{"puts":[{"type":"xml","ttlseconds":60,"key":"3e6713fb-1aea-4867-be7b-ffc75c8f74dd","value":"<?xml version=\"1.0\">__video_info__<\\/VAST>\r\n"}]}' http://localhost:2424/cache`
 
> Did you populate the cache again after each time you shut it down for the failure test ?


Yes. Everytime I ran the experiment, I repopulated under the `"3e6713fb-1aea-4867-be7b-ffc75c8f74dd"` using the `curl` command above. In order to simulate the server side errors I would simply turn off redis so the running Prebid Cache would get an error out of a non-running Redis backend.

Reply:
Hello @guscarreon , thanks a lot for the insights.
I agree with you about the `redis.Nil` defensive processing. I've applyed your suggestion and adapted the tests.